### PR TITLE
[test] Add test for gettxout to wallet.py

### PR DESCRIPTION
**Pull request by jimmysong**
Test gettxout as part of the wallet test.
Sanity-checks the utxo generated from coinbase.

### CODE REVIEW

**Comment by TheBlueMatt** [comment]
utACK. It would be cool to also test the include_mempool flag to gettxout to check that a) things which are spent in the mempool arent visible via gettxout (even though they are without include_mempool) and b) things which are only available in the mempool are inversely visible.

**Comment by jimmysong** [comment]
@TheBlueMatt Thank you for your excellent suggestion. I've made the changes as you suggested, squashed and rebased.

File: test/functional/wallet.py [hunk 1]
```diff
@@ -52,13 +52,33 @@ def run_test (self):
...
+        assert_equal(len(utxos), 1)
         assert_equal(len(self.nodes[1].listunspent()), 1)
         assert_equal(len(self.nodes[2].listunspent()), 0)
 
         # Send 21 BTC from 0 to 2 using sendtoaddress call.
         self.nodes[0].sendtoaddress(self.nodes[2].getnewaddress(), 11)
-        self.nodes[0].sendtoaddress(self.nodes[2].getnewaddress(), 10)
+        mempool_txid = self.nodes[0].sendtoaddress(self.nodes[2].getnewaddress(), 10)
+
+        # test gettxout
```
Comment (jnewbery) [inline test/functional/wallet.py]: nit: my personal preference is to make this kind of 'heading' comment into a log, so test runners have some feedback as they're running the test.

File: test/functional/wallet.py [hunk 2]
```diff
@@ -52,13 +52,33 @@ def run_test (self):
...
+        # confirmed tx should be visible if you exclude mempool
```
Comment (jnewbery) [inline test/functional/wallet.py]: This comment isn't quite right. We're not looking at a confirmed transaction here, we're looking at a transaction output who's spend is unconfirmed. That's why if include_mempool is set to False, we still think we own this output

File: test/functional/wallet.py [hunk 3]
```diff
@@ -52,13 +52,33 @@ def run_test (self):
...
+        # but invisible if you include mempool
+        confirmed_txid, confirmed_index = utxos[0]["txid"], utxos[0]["vout"]
+        txout = self.nodes[0].gettxout(confirmed_txid, confirmed_index, False)
+        assert_equal(txout['value'], 50)
+        txout = self.nodes[0].gettxout(confirmed_txid, confirmed_index, True)
+        assert(txout is None)
```
Comment (jnewbery) [inline test/functional/wallet.py]: no need for brackets here. [assert is a statement, not a function](http://stackoverflow.com/a/30770447/933705).

This should just be: `assert not txout`

File: test/functional/wallet.py [hunk 4]
```diff
@@ -52,13 +52,33 @@ def run_test (self):
...
+        # mempool tx should be invisible if you exclude mempool
+        # but visible if you include mempool
+        mempool_index = 0
+        txout = self.nodes[0].gettxout(mempool_txid, mempool_index, False)
+        assert(txout is None)
+        txout = self.nodes[0].gettxout(mempool_txid, mempool_index, True)
+        # note the mempool tx will have randomly assigned indices
+        # but 10 will go to node2 and the rest will go to node0
+        balance = self.nodes[0].getbalance()
+        assert(txout['value'] in (10, balance))
```
Comment (jnewbery) [inline test/functional/wallet.py]: I think the intent is clearer if you collect both txouts above, and then test that one of them is equal to 10 and the other is equal to balance. Something like:

```
txout0 = self.nodes[0].gettxout(mempool_txid, 0, True)
txout1 = self.nodes[0].gettxout(mempool_txid, 1, True)

balance = self.nodes[0].getbalance()
assert_equal(set(10, balance), set(txout0['value'], txout1['value']))
```

File: test/functional/wallet.py [hunk 5]
```diff
@@ -52,13 +52,33 @@ def run_test (self):
...
         # Send 21 BTC from 0 to 2 using sendtoaddress call.
         self.nodes[0].sendtoaddress(self.nodes[2].getnewaddress(), 11)
-        self.nodes[0].sendtoaddress(self.nodes[2].getnewaddress(), 10)
+        mempool_txid = self.nodes[0].sendtoaddress(self.nodes[2].getnewaddress(), 10)
+
+        # test gettxout
+        # confirmed tx should be visible if you exclude mempool
+        # but invisible if you include mempool
+        confirmed_txid, confirmed_index = utxos[0]["txid"], utxos[0]["vout"]
+        txout = self.nodes[0].gettxout(confirmed_txid, confirmed_index, False)
```
Comment (jnewbery) [inline test/functional/wallet.py]: consider using named arguments when the positional arguments are just bools. It makes it difficult for the reader to know what the argument is for (alright, I know you've explained it in the comment above, but the reader still needs to trust that you're right!)

**Review by jnewbery — COMMENTED** [review]
Great. Thanks for adding this.

A few nits, but otherwise looks good.

File: test/functional/wallet.py (hunk 4 above)
Comment (jimmysong) [inline test/functional/wallet.py]: good idea

**Comment by jimmysong** [comment]
nits addressed, rebased and squashed.

File: test/functional/wallet.py [hunk 6]
```diff
@@ -52,13 +52,33 @@ def run_test (self):
...
+        assert_equal(txout['value'], 50)
+        txout = self.nodes[0].gettxout(confirmed_txid, confirmed_index, True)
+        assert txout is None
+        # new utxo from mempool should be invisible if you exclude mempool
+        # but visible if you include mempool
+        txout = self.nodes[0].gettxout(mempool_txid, 0, False)
+        assert txout is None
+        txout1 = self.nodes[0].gettxout(mempool_txid, 0, True)
+        txout2 = self.nodes[0].gettxout(mempool_txid, 1, True)
```
Comment (MarcoFalke) [inline test/functional/wallet.py]: nit: In the future you might use named args for arguments that are unnamed boolean or integers of a function that takes more than one arg.

**Pull request merged**